File: middleware/python-lsp/src/failkit_lsp/code_actions.py
"""
Code Actions Module

Provides quick fixes and refactoring actions for F.A.I.L. Kit diagnostics.
"""

from typing import List, Optional

from lsprotocol.types import (
    CodeAction,
    CodeActionKind,
    Diagnostic,
    Position,
    Range,
    TextEdit,
    WorkspaceEdit,
)


def get_code_actions_for_diagnostic(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """
    Generate code actions for a specific diagnostic.
    
    Args:
        diagnostic: The diagnostic to generate actions for.
        document_uri: The URI of the document.
        document_text: The full text of the document.
    
    Returns:
        List of applicable CodeAction objects.
    """
    actions: List[CodeAction] = []
    rule_id = diagnostic.code
    
    if rule_id == "FK001":
        actions.extend(_get_receipt_actions(diagnostic, document_uri, document_text))
    elif rule_id == "FK002":
        actions.extend(_get_error_handling_actions(diagnostic, document_uri, document_text))
    elif rule_id in ("FK003", "FK007"):
        actions.extend(_get_secret_actions(diagnostic, document_uri, document_text))
    elif rule_id == "FK004":
        actions.extend(_get_confirmation_actions(diagnostic, document_uri, document_text))
    elif rule_id == "FK005":
        actions.extend(_get_resilience_actions(diagnostic, document_uri, document_text))
    elif rule_id == "FK009":
        actions.extend(_get_termination_actions(diagnostic, document_uri, document_text))
    
    # Always add a disable comment action
    actions.append(_get_disable_action(diagnostic, document_uri, document_text))
    
    return actions


def _get_receipt_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for missing receipt issues."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    indent = len(line) - len(line.lstrip())
    indent_str = " " * indent
    
    # Action 1: Add receipt generation after the call
    receipt_code = f'''
{indent_str}# F.A.I.L. Kit Receipt Generation
{indent_str}from datetime import datetime
{indent_str}import hashlib
{indent_str}import uuid
{indent_str}
{indent_str}receipt = {{
{indent_str}    "action_id": f"act_{{uuid.uuid4().hex[:8]}}",
{indent_str}    "timestamp": datetime.utcnow().isoformat() + "Z",
{indent_str}    "tool_name": "agent_execution",
{indent_str}    "status": "success",
{indent_str}    "input_hash": hashlib.sha256(str(input_data).encode()).hexdigest()[:16],
{indent_str}    "output_hash": hashlib.sha256(str(result).encode()).hexdigest()[:16],
{indent_str}}}
{indent_str}# TODO: Store receipt for audit trail
'''
    
    # Insert after the current line
    insert_position = Position(line=line_idx + 1, character=0)
    
    actions.append(CodeAction(
        title="Add receipt generation (F.A.I.L. Kit)",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(start=insert_position, end=insert_position),
                        new_text=receipt_code
                    )
                ]
            }
        ),
        is_preferred=True,
    ))
    
    # Action 2: Wrap with ReceiptGeneratingTool import suggestion
    actions.append(CodeAction(
        title="Add ReceiptGeneratingTool import",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=0, character=0),
                            end=Position(line=0, character=0)
                        ),
                        new_text="from fail_kit_langchain import ReceiptGeneratingTool, extract_receipts_from_agent_executor\n"
                    )
                ]
            }
        ),
    ))
    
    return actions


def _get_error_handling_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for missing error handling."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    indent = len(line) - len(line.lstrip())
    indent_str = " " * indent
    inner_indent = " " * (indent + 4)
    
    # Get the code on this line (the call that needs wrapping)
    call_code = line.strip()
    
    # Action: Wrap in try/except
    wrapped_code = f'''{indent_str}try:
{inner_indent}{call_code}
{indent_str}except Exception as e:
{inner_indent}# F.A.I.L. Kit: Handle agent/LLM failure
{inner_indent}import logging
{inner_indent}logging.error(f"Agent execution failed: {{e}}")
{inner_indent}raise  # Re-raise after logging
'''
    
    actions.append(CodeAction(
        title="Wrap in try/except (F.A.I.L. Kit)",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=line_idx, character=0),
                            end=Position(line=line_idx + 1, character=0)
                        ),
                        new_text=wrapped_code
                    )
                ]
            }
        ),
        is_preferred=True,
    ))
    
    return actions


def _get_secret_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for secret exposure issues."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    
    # Try to extract variable name and suggest env var replacement
    import re
    var_match = re.search(r"(\w+)\s*=\s*['\"]", line)
    if var_match:
        var_name = var_match.group(1)
        env_var_name = var_name.upper()
        
        # Replace with os.environ
        new_line = re.sub(
            r"['\"][^'\"]+['\"]",
            f'os.environ.get("{env_var_name}")',
            line
        )
        
        actions.append(CodeAction(
            title=f"Replace with os.environ.get('{env_var_name}')",
            kind=CodeActionKind.QuickFix,
            diagnostics=[diagnostic],
            edit=WorkspaceEdit(
                changes={
                    document_uri: [
                        TextEdit(
                            range=Range(
                                start=Position(line=line_idx, character=0),
                                end=Position(line=line_idx + 1, character=0)
                            ),
                            new_text=new_line + "\n"
                        )
                    ]
                }
            ),
            is_preferred=True,
        ))
    
    return actions


def _get_confirmation_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for side-effect confirmation issues."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    indent = len(line) - len(line.lstrip())
    indent_str = " " * indent
    inner_indent = " " * (indent + 4)
    
    call_code = line.strip()
    
    # Action: Add confirmation check
    confirmation_code = f'''{indent_str}# F.A.I.L. Kit: Require confirmation for destructive action
{indent_str}if not confirm_action("This will perform a destructive operation. Continue?"):
{inner_indent}raise ValueError("Operation cancelled by user")
{indent_str}{call_code}
'''
    
    actions.append(CodeAction(
        title="Add confirmation check (F.A.I.L. Kit)",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=line_idx, character=0),
                            end=Position(line=line_idx + 1, character=0)
                        ),
                        new_text=confirmation_code
                    )
                ]
            }
        ),
    ))
    
    return actions


def _get_resilience_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for missing resilience patterns."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    indent = len(line) - len(line.lstrip())
    indent_str = " " * indent
    
    # Action: Add retry decorator import and suggestion
    retry_import = "from tenacity import retry, stop_after_attempt, wait_exponential\n"
    
    actions.append(CodeAction(
        title="Add retry decorator import (tenacity)",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=0, character=0),
                            end=Position(line=0, character=0)
                        ),
                        new_text=retry_import
                    )
                ]
            }
        ),
    ))
    
    # Action: Add timeout configuration comment
    timeout_suggestion = f"{indent_str}# TODO: Add timeout parameter, e.g., timeout=30.0\n"
    
    actions.append(CodeAction(
        title="Add timeout TODO comment",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=line_idx, character=0),
                            end=Position(line=line_idx, character=0)
                        ),
                        new_text=timeout_suggestion
                    )
                ]
            }
        ),
    ))
    
    return actions


def _get_termination_actions(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> List[CodeAction]:
    """Generate code actions for missing termination configuration."""
    actions = []
    lines = document_text.splitlines()
    line_idx = diagnostic.range.start.line
    
    if line_idx >= len(lines):
        return actions
    
    line = lines[line_idx]
    
    # Find where to insert max_turns parameter
    import re
    
    # For initiate_chat calls
    if "initiate_chat" in line:
        # Add max_turns parameter
        if ")" in line:
            new_line = line.replace(")", ", max_turns=10)")
            actions.append(CodeAction(
                title="Add max_turns=10 parameter",
                kind=CodeActionKind.QuickFix,
                diagnostics=[diagnostic],
                edit=WorkspaceEdit(
                    changes={
                        document_uri: [
                            TextEdit(
                                range=Range(
                                    start=Position(line=line_idx, character=0),
                                    end=Position(line=line_idx + 1, character=0)
                                ),
                                new_text=new_line + "\n"
                            )
                        ]
                    }
                ),
                is_preferred=True,
            ))
    
    # For Agent configurations
    elif "Agent(" in line:
        if ")" in line:
            new_line = line.replace(")", ", max_consecutive_auto_reply=5)")
            actions.append(CodeAction(
                title="Add max_consecutive_auto_reply=5 parameter",
                kind=CodeActionKind.QuickFix,
                diagnostics=[diagnostic],
                edit=WorkspaceEdit(
                    changes={
                        document_uri: [
                            TextEdit(
                                range=Range(
                                    start=Position(line=line_idx, character=0),
                                    end=Position(line=line_idx + 1, character=0)
                                ),
                                new_text=new_line + "\n"
                            )
                        ]
                    }
                ),
                is_preferred=True,
            ))
    
    return actions


def _get_disable_action(
    diagnostic: Diagnostic,
    document_uri: str,
    document_text: str
) -> CodeAction:
    """Generate action to disable the diagnostic for this line."""
    line_idx = diagnostic.range.start.line
    lines = document_text.splitlines()
    
    if line_idx >= len(lines):
        indent_str = ""
    else:
        line = lines[line_idx]
        indent = len(line) - len(line.lstrip())
        indent_str = " " * indent
    
    disable_comment = f"{indent_str}# fail-kit-disable: {diagnostic.code}\n"
    
    return CodeAction(
        title=f"Disable {diagnostic.code} for this line",
        kind=CodeActionKind.QuickFix,
        diagnostics=[diagnostic],
        edit=WorkspaceEdit(
            changes={
                document_uri: [
                    TextEdit(
                        range=Range(
                            start=Position(line=line_idx, character=0),
                            end=Position(line=line_idx, character=0)
                        ),
                        new_text=disable_comment
                    )
                ]
            }
        ),
    )

File: vscode-extension/src/codeActions.ts
/**
 * F.A.I.L. Kit Code Action Provider
 *
 * Provides quick fixes for agent code issues detected by diagnostics.
 * Enhanced with dashboard links and severity indicators.
 */

import * as vscode from 'vscode';
import { DIAGNOSTIC_SOURCE } from './diagnostics';

/**
 * Extract rule ID from diagnostic code
 */
function getRuleIdFromDiagnostic(diagnostic: vscode.Diagnostic): string {
  if (typeof diagnostic.code === 'object' && diagnostic.code !== null) {
    return String(diagnostic.code.value);
  }
  return String(diagnostic.code);
}

/**
 * Quick fix templates for different issue types
 */
const QUICK_FIX_TEMPLATES = {
  FK001: {
    todo: '// TODO: [FAIL-Kit] Add receipt generation for audit compliance\n',
    receiptSnippet: `
// Generate receipt for audit trail
const receipt = {
  action_id: \`act_\${crypto.randomBytes(4).toString('hex')}\`,
  tool_name: '$TOOL_NAME',
  timestamp: new Date().toISOString(),
  status: 'success',
  input_hash: hashData(input),
  output_hash: hashData(result),
};
`,
    wrapperSnippet: `
// Wrap with receipt generation
const receiptResult = await generateReceiptForAction(async () => {
  $ORIGINAL_CODE
});
`,
  },
  FK002: {
    todo: '// TODO: [FAIL-Kit] Add error handling for this LLM call\n',
    tryCatchSnippet: `try {
  $ORIGINAL_CODE
} catch (error) {
  console.error('LLM call failed:', error);
  throw error;
}
`,
  },
  FK003: {
    todo: '// TODO: [FAIL-Kit] Move this secret to environment variable\n',
  },
  FK004: {
    todo: '// TODO: [FAIL-Kit] Add confirmation before this destructive operation\n',
  },
  FK005: {
    todo: '// TODO: [FAIL-Kit] Add timeout and retry logic for this LLM call\n',
  },
  FK006: {
    todo: '// TODO: [FAIL-Kit] Add action_id and timestamp for provenance\n',
  },
  FK007: {
    todo: '// TODO: [FAIL-Kit] Remove hardcoded credential - use environment variable\n',
  },
};

/**
 * Code action provider for F.A.I.L. Kit quick fixes
 */
export class FailKitCodeActionProvider implements vscode.CodeActionProvider {
  public static readonly providedCodeActionKinds = [
    vscode.CodeActionKind.QuickFix,
    vscode.CodeActionKind.Refactor,
  ];

  /**
   * Provide code actions for diagnostics
   */
  provideCodeActions(
    document: vscode.TextDocument,
    range: vscode.Range | vscode.Selection,
    context: vscode.CodeActionContext,
    _token: vscode.CancellationToken
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    for (const diagnostic of context.diagnostics) {
      if (diagnostic.source !== DIAGNOSTIC_SOURCE) {
        continue;
      }

      const ruleId = getRuleIdFromDiagnostic(diagnostic);

      // Add "View in Dashboard" action for all rules
      actions.push(this.createDashboardAction(document, diagnostic, ruleId));

      // Add "Auto-Fix" action for supported rules
      if (['FK001', 'FK002', 'FK003', 'FK004', 'FK005', 'FK006', 'FK007'].includes(ruleId)) {
        actions.push(this.createAutoFixAction(document, diagnostic, ruleId));
      }

      switch (ruleId) {
        case 'FK001':
          actions.push(...this.createFK001Actions(document, diagnostic));
          break;
        case 'FK002':
          actions.push(...this.createFK002Actions(document, diagnostic));
          break;
        case 'FK003':
        case 'FK007':
          actions.push(...this.createSecretActions(document, diagnostic, ruleId));
          break;
        case 'FK004':
          actions.push(...this.createSideEffectActions(document, diagnostic));
          break;
        case 'FK005':
          actions.push(...this.createResilienceActions(document, diagnostic));
          break;
        case 'FK006':
          actions.push(...this.createProvenanceActions(document, diagnostic));
          break;
        default:
          // Generic TODO action for unknown rules
          actions.push(this.createTodoAction(document, diagnostic, ruleId));
          break;
      }
    }

    return actions;
  }

  /**
   * Create "View in Dashboard" action
   */
  private createDashboardAction(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic,
    ruleId: string
  ): vscode.CodeAction {
    const action = new vscode.CodeAction(
      '$(dashboard) View in Dashboard',
      vscode.CodeActionKind.QuickFix
    );
    action.command = {
      command: 'fail-kit.showInDashboard',
      title: 'View in Dashboard',
      arguments: [document.uri.fsPath, diagnostic.range.start.line, ruleId],
    };
    return action;
  }

  /**
   * Create "Auto-Fix" action
   */
  private createAutoFixAction(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic,
    ruleId: string
  ): vscode.CodeAction {
    const action = new vscode.CodeAction(
      '$(wand) Auto-Fix This Issue',
      vscode.CodeActionKind.QuickFix
    );
    action.command = {
      command: 'fail-kit.autoFixIssue',
      title: 'Auto-Fix Issue',
      arguments: [document.uri.fsPath, diagnostic.range.start.line, ruleId],
    };
    action.isPreferred = true;
    return action;
  }

  /**
   * Create actions for FK003/FK007: Secret exposure
   */
  private createSecretActions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic,
    ruleId: string
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];
    const template = ruleId === 'FK003' ? QUICK_FIX_TEMPLATES.FK003 : QUICK_FIX_TEMPLATES.FK007;

    // Action 1: Add TODO comment
    const todoAction = new vscode.CodeAction(
      'Add TODO: Move secret to environment variable',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) + template.todo
    );
    todoAction.diagnostics = [diagnostic];
    actions.push(todoAction);

    // Action 2: Show security documentation
    const docsAction = new vscode.CodeAction(
      'Learn about secure secret management',
      vscode.CodeActionKind.QuickFix
    );
    docsAction.command = {
      command: 'fail-kit.showDocumentation',
      title: 'Show Documentation',
      arguments: ['secrets'],
    };
    actions.push(docsAction);

    return actions;
  }

  /**
   * Create actions for FK004: Side-effect without confirmation
   */
  private createSideEffectActions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    const todoAction = new vscode.CodeAction(
      'Add TODO: Add confirmation for destructive operation',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) + QUICK_FIX_TEMPLATES.FK004.todo
    );
    todoAction.diagnostics = [diagnostic];
    actions.push(todoAction);

    // Disable action
    actions.push(this.createDisableAction(document, diagnostic));

    return actions;
  }

  /**
   * Create actions for FK005: LLM without resilience
   */
  private createResilienceActions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    const todoAction = new vscode.CodeAction(
      'Add TODO: Add timeout/retry for LLM call',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) + QUICK_FIX_TEMPLATES.FK005.todo
    );
    todoAction.diagnostics = [diagnostic];
    actions.push(todoAction);

    actions.push(this.createDisableAction(document, diagnostic));

    return actions;
  }

  /**
   * Create actions for FK006: Missing provenance
   */
  private createProvenanceActions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    const todoAction = new vscode.CodeAction(
      'Add TODO: Add provenance metadata',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) + QUICK_FIX_TEMPLATES.FK006.todo
    );
    todoAction.diagnostics = [diagnostic];
    actions.push(todoAction);

    actions.push(this.createDisableAction(document, diagnostic));

    return actions;
  }

  /**
   * Create disable action for any rule
   */
  private createDisableAction(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction {
    const disableAction = new vscode.CodeAction(
      'Disable F.A.I.L. Kit for this line',
      vscode.CodeActionKind.QuickFix
    );
    disableAction.edit = new vscode.WorkspaceEdit();
    disableAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        '// fail-kit-disable-next-line\n'
    );
    disableAction.diagnostics = [diagnostic];
    return disableAction;
  }

  /**
   * Create actions for FK001: Missing receipt
   */
  private createFK001Actions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    // Action 1: Add TODO comment
    const todoAction = new vscode.CodeAction(
      'Add TODO: Generate receipt for this tool call',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        QUICK_FIX_TEMPLATES.FK001.todo
    );
    todoAction.diagnostics = [diagnostic];
    todoAction.isPreferred = true;
    actions.push(todoAction);

    // Action 2: Add fail-kit-disable comment
    const disableAction = new vscode.CodeAction(
      'Disable F.A.I.L. Kit for this line',
      vscode.CodeActionKind.QuickFix
    );
    disableAction.edit = new vscode.WorkspaceEdit();
    disableAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        '// fail-kit-disable-next-line\n'
    );
    disableAction.diagnostics = [diagnostic];
    actions.push(disableAction);

    // Action 3: Show documentation
    const docsAction = new vscode.CodeAction(
      'Learn about receipt generation',
      vscode.CodeActionKind.QuickFix
    );
    docsAction.command = {
      command: 'fail-kit.showDocumentation',
      title: 'Show Documentation',
      arguments: ['receipts'],
    };
    actions.push(docsAction);

    return actions;
  }

  /**
   * Create actions for FK002: Missing error handling
   */
  private createFK002Actions(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction[] {
    const actions: vscode.CodeAction[] = [];

    // Action 1: Add TODO comment
    const todoAction = new vscode.CodeAction(
      'Add TODO: Add error handling for LLM call',
      vscode.CodeActionKind.QuickFix
    );
    todoAction.edit = new vscode.WorkspaceEdit();
    todoAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        QUICK_FIX_TEMPLATES.FK002.todo
    );
    todoAction.diagnostics = [diagnostic];
    todoAction.isPreferred = true;
    actions.push(todoAction);

    // Action 2: Wrap in try-catch (using snippet)
    const wrapAction = new vscode.CodeAction(
      'Wrap in try-catch block',
      vscode.CodeActionKind.Refactor
    );
    wrapAction.command = {
      command: 'fail-kit.wrapInTryCatch',
      title: 'Wrap in try-catch',
      arguments: [document.uri, diagnostic.range],
    };
    actions.push(wrapAction);

    // Action 3: Add fail-kit-disable comment
    const disableAction = new vscode.CodeAction(
      'Disable F.A.I.L. Kit for this line',
      vscode.CodeActionKind.QuickFix
    );
    disableAction.edit = new vscode.WorkspaceEdit();
    disableAction.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        '// fail-kit-disable-next-line\n'
    );
    disableAction.diagnostics = [diagnostic];
    actions.push(disableAction);

    return actions;
  }

  /**
   * Create a generic TODO action for any rule
   */
  private createTodoAction(
    document: vscode.TextDocument,
    diagnostic: vscode.Diagnostic,
    ruleId: string
  ): vscode.CodeAction {
    const action = new vscode.CodeAction(
      `Add TODO: Fix ${ruleId} issue`,
      vscode.CodeActionKind.QuickFix
    );
    action.edit = new vscode.WorkspaceEdit();
    action.edit.insert(
      document.uri,
      new vscode.Position(diagnostic.range.start.line, 0),
      this.getIndentation(document, diagnostic.range.start.line) +
        `// TODO: [FAIL-Kit ${ruleId}] ${diagnostic.message}\n`
    );
    action.diagnostics = [diagnostic];
    return action;
  }

  /**
   * Get the indentation of a line
   */
  private getIndentation(document: vscode.TextDocument, lineNumber: number): string {
    const line = document.lineAt(lineNumber);
    const match = line.text.match(/^(\s*)/);
    return match ? match[1] : '';
  }
}

/**
 * Register commands for code actions
 */
export function registerCodeActionCommands(
  context: vscode.ExtensionContext
): void {
  // Command: Wrap in try-catch
  context.subscriptions.push(
    vscode.commands.registerCommand(
      'fail-kit.wrapInTryCatch',
      async (uri: vscode.Uri, range: vscode.Range) => {
        const document = await vscode.workspace.openTextDocument(uri);
        const editor = await vscode.window.showTextDocument(document);

        // Get the line content
        const lineText = document.lineAt(range.start.line).text;
        const indent = lineText.match(/^(\s*)/)?.[1] || '';

        // Create try-catch wrapper
        const wrappedCode = `${indent}try {\n${lineText}\n${indent}} catch (error) {\n${indent}  console.error('Operation failed:', error);\n${indent}  throw error;\n${indent}}\n`;

        // Replace the line
        await editor.edit((editBuilder) => {
          const fullLineRange = document.lineAt(range.start.line).range;
          editBuilder.replace(fullLineRange, wrappedCode);
        });
      }
    )
  );

  // Command: Show documentation
  context.subscriptions.push(
    vscode.commands.registerCommand('fail-kit.showDocumentation', (topic: string) => {
      const baseUrl = 'https://github.com/resetroot99/The-FAIL-Kit';
      let url = baseUrl;

      switch (topic) {
        case 'receipts':
          url = `${baseUrl}/blob/main/receipt-standard/README.md`;
          break;
        case 'error-handling':
          url = `${baseUrl}/blob/main/docs/SECURITY_TESTING.md`;
          break;
        default:
          url = `${baseUrl}#readme`;
      }

      vscode.env.openExternal(vscode.Uri.parse(url));
    })
  );
}
